### PR TITLE
RSDK-6972: Add Shutdown Command to Robot

### PR DESCRIPTION
**Pull request by danielbotros**
### Description
**RSDK-6972:** Add a Shutdown command to robot
Scope doc can be found [here](https://docs.google.com/document/d/1LR-fnNpy47qAGR0flFqDCe6sJUzB9n6E9lrG0XyJdaI/edit#heading=h.rw9apvrdt84j)
* `Shutdown` command, `ShutdownRequest`, and `ShutdownResponse` added to API in this [PR](https://github.com/viamrobotics/api/commit/4f9d14d067b9b7965fcb4e89d6185e9dc348c3cd)
* Ability to call `Shutdown()` added through Robot interface, client, and server. 
* A robot cannot shut itself down, it must be shutdown by its entry point process. In order for the robot to be able to shut itself down, it must have access to a callback function which allows it to call `cancel()` on it's original context. 
  * In order to accomplish this, a shutdown function (log of stack trace + `cancel()`) is binded to the robot options by adding a `shutdownCallback` field to the  robot options struct. This is set through `WithShutdownCallback()` in `entrypoint.go`
  * Robots now have a `shutdownCallback` field which is set during initialization through robot options. On `Shutdown()`, this callback is called to shutdown the robot, if it exists, otherwise it throws an error.  
    * May not exist since it is only binded when being started through `entrypoint.go `

### Testing
* A single integration test is added to `entrypoint_test.go` which starts the RDK as a separate process then calls `Shutdown()` and ensures it does not return an error and that there are no errors in the logs as the shutdown occurs. 
* Tests are added to `robot/server/server_test.go` and `robot/client/client_test.go` for server and client calls to `Shutdown()` respectively

### CODE REVIEW

File: robot/client/client.go [hunk 1]
```diff
@@ -990,3 +990,10 @@ func (rc *RobotClient) RestartModule(ctx context.Context, req robot.RestartModul
 	}
 	return nil
 }
+
+// Shutdown shuts down the robot.
+func (rc *RobotClient) Shutdown(ctx context.Context) error {
```
Comment (cheukt) [inline robot/client/client.go:995]: can you add a test for this as well? add it through updating `injectRobot`

File: robot/server/server.go [hunk 2]
```diff
@@ -478,3 +478,12 @@ func (s *Server) RestartModule(ctx context.Context, req *pb.RestartModuleRequest
 	}
 	return &pb.RestartModuleResponse{}, nil
 }
+
+// Shutdown shuts down the robot.
+func (s *Server) Shutdown(ctx context.Context, _ *pb.ShutdownRequest) (*pb.ShutdownResponse, error) {
+	err := s.robot.Shutdown(ctx)
```
Comment (cheukt) [inline robot/server/server.go:484]: add a test through updating `injectRobot`

File: web/server/entrypoint.go [hunk 3]
```diff
@@ -359,6 +359,19 @@ func (s *robotServer) serveWeb(ctx context.Context, cfg *config.Config) (err err
 		robotOptions = append(robotOptions, robotimpl.WithRevealSensitiveConfigDiffs())
 	}
 
+	shutdownCallbackOpt := robotimpl.WithShutdownCallback(func() {
+		bufSize := 1 << 20
```
Comment (cheukt) [inline web/server/entrypoint.go:363]: factor this out so we aren't copying the code from https://github.com/viamrobotics/rdk/pull/4032/files#diff-780f3d307fb668fed30f86c777b47efbbd6848b5aa3282dc1958bc76b2f07ecbR343-R350

File: web/server/entrypoint_test.go [hunk 4]
```diff
@@ -106,4 +107,49 @@ func TestEntrypoint(t *testing.T) {
 			test.That(t, reg.Schema, test.ShouldNotBeNil)
 		}
 	})
+	t.Run("shutdown functionality", func(t *testing.T) {
```
Comment (cheukt) [inline web/server/entrypoint_test.go:110]: make this a separate test

**Review by cheukt — CHANGES_REQUESTED** [review]
looks good, some requests around better structure and additional tests

File: robot/impl/local_robot.go [hunk 5]
```diff
@@ -1301,3 +1303,12 @@ func (r *localRobot) RestartModule(ctx context.Context, req robot.RestartModuleR
...
 	return nil
 }
+
+func (r *localRobot) Shutdown(ctx context.Context) error {
+	shutdownFunc := r.shutdownCallback
+	if shutdownFunc != nil {
+		shutdownFunc()
+		return nil
+	}
+	return errors.New("shutdown function not defined")
```
[the remark this answers is not in the document]
Comment (maximpertsov) [reply robot/impl/local_robot.go:1313]: error is fine here - we should just document the behavior. we could also consider validating/requiring the shutdown "option" at an earlier step, but i think what you have is perfectly fine.

as an aside, long-term we should consider merging the server management logic into the robot itself... it's pretty confusing that robot has both a `Shutdown` and `Close` method now, where the former indirectly calls the latter at some point.

[the remark this answers is not in the document]
Comment (danielbotros) [reply robot/impl/local_robot.go:1313]: added some more documentation outlining the command will throw an error if a shutdown callback isn't binded to the robot

I definitely agree, it makes sense why both exist but it is not intuitive at all on how/which to use as a user

File: web/server/entrypoint_test.go (hunk 4 above)
Comment (danielbotros) [reply web/server/entrypoint_test.go:116]: all donezo

File: robot/client/client.go [hunk 6]
```diff
@@ -990,3 +990,11 @@ func (rc *RobotClient) RestartModule(ctx context.Context, req robot.RestartModul
 	}
 	return nil
 }
+
+// Shutdown shuts down the robot. Will throw a "shutdown function not defined error" if
+// shutdown callback is not binded to robot options.
```
Comment (benjirewis) [inline robot/client/client.go:995]:
```suggestion
// Shutdown shuts down the robot. Will return a "shutdown function not defined error" if
// shutdown callback is not specified in robot options.
```
Nits.

File: web/server/entrypoint_test.go [hunk 7]
```diff
@@ -107,3 +108,54 @@ func TestEntrypoint(t *testing.T) {
...
+		conn, err := robottestutils.Connect(port)
+		test.That(t, err, test.ShouldBeNil)
+		defer func() {
+			test.That(t, conn.Close(), test.ShouldBeNil)
+		}()
+		rc := robotpb.NewRobotServiceClient(conn)
+
+		_, err = rc.Shutdown(context.Background(), &robotpb.ShutdownRequest{})
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, logObserver.FilterLevelExact(zapcore.ErrorLevel).Len(), test.ShouldEqual, 0)
```
Comment (benjirewis) [inline web/server/entrypoint_test.go:159]: Nice 🧑‍🔧 .

**Review by benjirewis — APPROVED** [review]
Looks good to me! 🫡  Thanks

File: robot/client/client_test.go [hunk 8]
```diff
@@ -2044,3 +2044,27 @@ func TestCloudMetadata(t *testing.T) {
...
+	defer gServer.Stop()
+
+	client, err := New(context.Background(), listener.Addr().String(), logger)
+	test.That(t, err, test.ShouldBeNil)
+	defer func() {
+		test.That(t, client.Close(context.Background()), test.ShouldBeNil)
+	}()
+
+	err = client.Shutdown(context.Background())
+	test.That(t, err, test.ShouldBeNil)
```
Comment (cheukt) [inline robot/client/client_test.go:2069]: this doesn't properly test that the client sends a message to the server - we should do something like https://github.com/viamrobotics/rdk/blob/main/components/arm/server_test.go#L84-L87 and maybe toggle a bool to indicate we reached the Shutdown func

File: robot/server/server_test.go [hunk 9]
```diff
@@ -304,6 +304,18 @@ func TestServer(t *testing.T) {
...
+
+	t.Run("Shutdown", func(t *testing.T) {
+		injectRobot := &inject.Robot{}
+		injectRobot.ResourceRPCAPIsFunc = func() []resource.RPCAPI { return nil }
+		injectRobot.ResourceNamesFunc = func() []resource.Name { return nil }
+		injectRobot.ShutdownFunc = func(ctx context.Context) error { return nil }
+		server := server.New(injectRobot)
+		req := pb.ShutdownRequest{}
+
+		_, err := server.Shutdown(context.Background(), &req)
```
Comment (cheukt) [inline robot/server/server_test.go:316]: same as above

**Review by cheukt — CHANGES_REQUESTED** [review]
small change in tests requested

File: robot/client/client.go (hunk 6 above)
Comment (danielbotros) [reply robot/client/client.go]: updated doc

**Review by cheukt — APPROVED** [review]
LGTM!

File: robot/impl/local_robot.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (maximpertsov) [reply robot/impl/local_robot.go:1313]: doc looks good!

**Review by maximpertsov — APPROVED** [review]

**Pull request merged**